tokio-util/src/codec/bytes_codec.rs

```diff
@@ -65,11 +65,14 @@ impl Decoder for BytesCodec {
     }
 }
 
-impl Encoder for BytesCodec {
-    type Item = Bytes;
+impl<T> Encoder<T> for BytesCodec
+where
+    T: AsRef<[u8]>,
```

Comment:
Should this be `Buf` instead of `AsRef`? I think if we use the AsRef type we always need to copy while if we take a `Buf` we can use `Buf::bytes` to avoid a copy I think.

Reply:
No idea what you mean, `as_ref` and `bytes` both return a `&[u8]` (which does not copy), however that slice do need to be copied regardless so the result would be equivalent. `AsRef` is more general however.

Reply:
This is using `BufMut::put()`, which I don't think specializes on `Bytes` instances. It can take any `Buf`, but of course `Buf` is a more powerful and thus more restrictive interface to implement than `AsRef<[u8]>` (even ignoring the difference between a `bytes` trait and a `std` trait).

Reply:
Ah I was thinking of this API https://docs.rs/bytes/0.5.4/bytes/buf/trait.Buf.html#method.to_bytes my mistake. More so I just don't see the advantage of using `AsRef` when we could just use `Buf`? Buf would allow us to do more specialized things.

I think if we change this to be `T: Buf` it would be a much better API than `AsRef`?

Reply:
Why would it be a better API? It makes no sense to me. As I understand it, `Buf` conceptually represents a cursor with backing memory. For the `Encoder`, we really just need a chunk of memory that gets appended to the `BytesMut`. `AsRef<[u8]>` is a better default since it's from std, so more things will implement it by default, and it's a smaller trait so it can be implemented by more things.

Reply:
I would think more things can implement `T: Buf` than the other way around.

Reply:
Why? That's not intuitive to me.

Reply:
Reverted to `Bytes` for now.